test/libmpv_test.c

```diff
@@ -15,15 +15,260 @@
  * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <inttypes.h>
 #include <libmpv/client.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
 
-int main(void)
+// Stolen from osdep/compiler.h
+#ifdef __GNUC__
+#define PRINTF_ATTRIBUTE(a1, a2) __attribute__ ((format(printf, a1, a2)))
+#define MP_NORETURN __attribute__((noreturn))
+#else
+#define PRINTF_ATTRIBUTE(a1, a2)
+#define MP_NORETURN
+#endif
+
+// Broken crap with __USE_MINGW_ANSI_STDIO
+#if defined(__MINGW32__) && defined(__GNUC__) && !defined(__clang__)
+#undef PRINTF_ATTRIBUTE
+#define PRINTF_ATTRIBUTE(a1, a2) __attribute__ ((format (gnu_printf, a1, a2)))
+#endif
+
+// Dummy values for the test.
+static const char *str = "string";
+static int flag = 1;
+static int64_t int_ = 20;
+static double double_ = 1.5;
+
+// Global handle.
+static mpv_handle *ctx;
+
+
+MP_NORETURN PRINTF_ATTRIBUTE(1, 2)
+static void fail(const char *fmt, ...)
 {
-    mpv_handle *handle = mpv_create();
-    if (!handle)
+	if (fmt) {
+		va_list va;
+		va_start(va, fmt);
+		vfprintf(stderr, fmt, va);
+		va_end(va);
+	}
+	mpv_destroy(ctx);
+    exit(1);
+}
+
+static void check_api_error(int status)
+{
+    if (status < 0)
+        fail("mpv API error: %s\n", mpv_error_string(status));
+}
+
+static void check_double(const char *property)
+{
+    double result_double;
+    check_api_error(mpv_get_property(ctx, property, MPV_FORMAT_DOUBLE, &result_double));
+    if (double_ != result_double)
+        fail("Double: expected '%f' but got '%f'!\n", double_, result_double);
+}
+
+static void check_flag(const char *property)
+{
+    int result_flag;
+    check_api_error(mpv_get_property(ctx, property, MPV_FORMAT_FLAG, &result_flag));
+    if (flag != result_flag)
+        fail("Flag: expected '%d' but got '%d'!\n", flag, result_flag);
+}
+
+static void check_int(const char *property)
+{
+    int64_t result_int;
+    check_api_error(mpv_get_property(ctx, property, MPV_FORMAT_INT64, &result_int));
+    if (int_ != result_int)
+        fail("Int: expected '%" PRId64 "' but got '%" PRId64 "'!\n", int_, result_int);
+}
+
+static void check_string(const char *property)
+{
+    char *result_string;
+    check_api_error(mpv_get_property(ctx, property, MPV_FORMAT_STRING, &result_string));
+    if (strcmp(str, result_string) != 0) {
+        // Don't immediately fail in order to free the string.
```

Comment:
Forgot to comment on this earlier, but I don't think this is needed since OS cleans up memory at exit.

If you're worried about some warning from some run-time leak-checker, it still shouldn't matter since the test *already* failed. So an additional warning form some leak-checker isn't going matter at that point.

So unless there's some static analyzer that's causing warning, I'd just `fail` directly.

Reply:
Actually, ASAN doesn't complain about this at all (!?) when I change some stuff around so it should leak. Not sure if that makes sense.

Reply:
> Actually, ASAN doesn't complain about this at all (!?) when I change some stuff around so it should leak. Not sure if that makes sense.

I can't tell about other leak checkers, but LeakSanitizer is usually "smart" not to flag things which are still "reachable" as leaks. So maybe that pointer is still reachable from somewhere which is why ASan doesn't flag it?

Does it complain if you remove both the `mpv_free` calls? What happens if you `strdup` the string and leak the dupped string intentionally?

Reply:
It will report the leak if I rig the test to exit gracefully, but if I let it go to `fail` it does not say anything. That includes using `strdup` and remove the `mpv_free` calls.

Reply:
> It will report the leak if I rig the test to exit gracefully, but if I let it go to fail it does not say anything.

Probably because it sees that those pointers are still reachable from the previous stack-frame when you call `fail()`.